internal/http/parser/http_test.go

```diff
@@ -0,0 +1,236 @@
+
+func TestParseResponseProperContent(t *testing.T) {
+	response, n, err := Parse(properContent)
+	require.NoError(t, err)
+	assert.Equal(t, n, len(properContent))
+
+	assert.Equal(t, "HTTP/1.1", response.StatusLine.Version)
+
+	assert.Equal(t, 200, response.StatusLine.StatusCode)
+	assert.Equal(t, "OK", response.StatusLine.Reason)
+
+	assert.Equal(t, "Werkzeug/3.0.2 Python/3.10.13", response.FindHeader("Server"))
+	assert.Equal(t, "Tue, 30 Apr 2024 06:16:31 GMT", response.FindHeader("Date"))
+	assert.Equal(t, "29", response.FindHeader("Content-Length"))
+
+	assert.Equal(t, "This is a plain text response", string(response.Body))
+}
+
+var protocolErrors = map[string]string{
+	"HTTP/3.0 200 OK": `
+Received: "HTTP/3.0 200 OK"
+                ^ error
+Error: Expected HTTP-version 1.X, Received: 3.0`,
+	"HTPP/1.0 200 OK": `
+Received: "HTPP/1.0 200 OK"
+               ^ error
+Error: Expected HTTP-version field to start with 'HTTP/'`,
+	"HTTP|1.0 200 OK": `
+Received: "HTTP|1.0 200 OK"
+               ^ error
+Error: Expected HTTP-version field to start with 'HTTP/'`,
+	"HTTP//1.0 200 OK": `
+Received: "HTTP//1.0 200 OK"
+                   ^ error
+Error: Invalid HTTP-version field length`,
+	"HTTP/1 200 OK": `
+Received: "HTTP/1 200 OK"
+                ^ error
+Error: Invalid HTTP-version field length`,
+	"HTTP/1.0 ZOO OK": `
+Received: "HTTP/1.0 ZOO OK"
+                      ^ error
+Error: Invalid status-code field, Expected integer value, Received: ZOO`,
+	"HTTP/1.0 2000 OK": `
+Received: "HTTP/1.0 2000 OK"
+                       ^ error
+Error: Invalid status-code field length, Expected: 3 digits, Received: 4`,
+	"HTTP/1.1 200 OK\r\nConnection : close\r\n": `
+Received: "200 OK\r\nConnection : close\r\n"
+                                 ^ error
+Error: No whitespace allowed before header separator`,
+}
+
+func TestParseVersion(t *testing.T) {
+	for response, error := range protocolErrors {
+		t.Run(response, func(t *testing.T) {
+			_, _, err := Parse([]byte(response))
+			assert.Error(t, err)
+			assert.IsType(t, BadProtocolError{}, err)
+			assert.EqualError(t, err, strings.TrimSpace(error))
+		})
+	}
+}
+
+var incompleteErrors = map[string]string{
+	"HTTP/1.0_200_OK": `
+Received: "HTTP/1.0_200_OK"
+                          ^ error
+Error: Expected: HTTP-version, Received: "HTTP/1.0_200_OK"`,
+	"HTTP/1.0 200 OK": `
+Received: "HTTP/1.0 200 OK"
+                          ^ error
+Error: Expected CRLF after status line`,
+	"HTTP/1.1 200 OK\r\nConnection: close": `
+Received: "K\r\nConnection: close"
+                                 ^ error
+Error: Expected CRLF after header value`,
+	"HTTP/1.1 200 OK\r\nConnection: close\r\n": `
+Received: "\nConnection: close\r\n"
+                                  ^ error
+Error: Expected empty line after header section`,
```

Comment:
@ryan-gang what does empty line here mean? Can we just say Expected `\r\n` instead?